BTSSS Claim Details Page

hinzed1127 opened the pull request:
## Are you removing, renaming or moving a folder in this PR?
- [x] No, I'm not changing any folders (skip to Summary and delete the rest of this section)
- [ ] Yes, I'm removing, renaming or moving a folder

If the folder you changed contains a `manifest.json`, search for its `entryName` in the content-build [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json) (the `entryName` there will match).

If an entry for this folder exists in content-build and you are:
1. **Deleting a folder**:
   1. First search `vets-website` for _all_ instances of the `entryName` in your `manifest.json` and remove them in a separate PR. Look particularly for references in `src/applications/static-pages/static-pages-entry.js` and `src/platform/forms/constants.js`. _**If you do not do this, other applications will break!**_
      - _Add the link to your merged vets-website PR here_
   2. Then, Delete the application entry in [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json) and merge that PR **before** this one
      - _Add the link to your merged content-build PR here_

2. **Renaming or moving a folder**: Update the entry in the [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json), but do not merge it until your vets-website changes here are merged. The content-build PR must be merged immediately after your vets-website change is merged in to avoid CI errors with content-build (and Tugboat).

### :warning: TeamSites :warning:
Examples of a TeamSite: https://va.gov/health and https://benefits.va.gov/benefits/. This scenario is also referred to as the "injected" header and footer. You can reach out in the `#sitewide-public-websites` Slack channel for questions.

Did you change site-wide styles, platform utilities or other infrastructure?
- [x] No
- [ ] Yes, and I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#that-sounds-normal-so-whats-the-proxy-all-about) to test the injected header scenario

## Summary

- Adds an MVP page for claim details, accessible via either:
	- linked on a status card in the travel claims list page
	- direct link via claim ID

## Related issue(s)

- https://github.com/department-of-veterans-affairs/va.gov-team/issues/92295
	- [Zenhub link of same issue](https://app.zenhub.com/workspaces/beneficiary-travel-btsss-65147e21930cd900223d8e64/issues/gh/department-of-veterans-affairs/va.gov-team/92295)
## Testing done

- added unit tests for new feature flags on status page
- added unit tests for details page
- added cypress test for navigating `claim status page -> details page -> status page` via client-side routing

## Screenshots

### New page

![Claim Details Page](https://github.com/user-attachments/assets/489e6748-6ebc-43e2-a724-0b22d2850c35)

### Updates to claim status list page

|         | Before | After |
| ------- | ------ | ----- |
| Mobile  |        |       |
| Desktop |   <img width="590" alt="Screenshot 2024-09-13 at 4 30 13 PM" src="https://github.com/user-attachments/assets/8a8c982a-d0ff-4709-859f-68fcf8054136">   |  <img width="593" alt="Screenshot 2024-09-13 at 4 27 31 PM" src="https://github.com/user-attachments/assets/71e00631-32a4-4229-af94-6b16d9844acb">  |

## What areas of the site does it impact?

Travel Pay
## Acceptance criteria

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Linting warnings have been addressed
- [x] Documentation has been updated ([link to documentation](#) \*if necessary)
- [x] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [ ] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [x] Did you login to a local build and verify all authenticated routes work as expected with a test user

## Requested Feedback

(OPTIONAL) _What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

hinzed1127 committed:
WIP Claim Details: routing, render details w/o fetching, etc

hinzed1127 committed:
Incorporate data fetching into details component

hinzed1127 committed:
WIP close to matching design spec

kjduensing commented on src/applications/travel-pay/components/TravelClaimDetails.jsx:
```diff
@@ -0,0 +1,137 @@
+        {claimDetails && (
+          <>
+            <h1>Your travel reimbursement claim for {appointmentDate}</h1>
+            <h2>Claim number: {claimNumber}</h2>
```
This variable name is correct, but the screenshot shows a UUID-formatted claim ID. This is not a useful ID to the Veteran as the UUID is an internal-to-btsss id. The claim number should have the format `TC[0-9]{13}`, eg `TC0123456789123`. I suspect a mock data issue, but something to look into.

kjduensing commented on src/applications/travel-pay/components/Breadcrumbs.jsx:
```diff
@@ -1,28 +1,71 @@
+      label: 'Claim Details',
+      isRouterLink: true,
+    });
+    // breadcrumbList = [
+    //   {
+    //     href: '/',
+    //     label: 'Back to your travel reimbursement claims',
+    //     isRouterLink: true,
+    //   },
+    // ];
```
Can we remove this commented code?

hinzed1127 replied:
see previous comment on clean-up

kjduensing commented on src/applications/travel-pay/components/Breadcrumbs.jsx:
```diff
@@ -1,28 +1,71 @@
-          label: 'Check your travel reimbursement claim status',
-        },
-      ]}
+      // homeVeteransAffairs={false}
```
Can we remove this commented code?

kjduensing commented on src/applications/travel-pay/components/TravelClaimCard.jsx:
```diff
@@ -50,6 +50,15 @@ export default function TravelClaimCard(props) {
           Updated on {updateDate} at {updateTime}
         </li>
       </ul>
+      <Link
+        to={{
+          pathname: `/${id}`,
+          state: { claimDetailsProps: props },
+        }}
+        className="vads-u-display--flex vads-u-align-items--center"
+      >
+        Travel reimbursement claim details <va-icon icon="chevron_right" />
+      </Link>
```
Let's put this behind the [details feature flag](https://github.com/department-of-veterans-affairs/vets-api/pull/18246/files). Since the status page exists in prod already, I don't want to put something new and untested out there before we're ready.

kjduensing commented on src/applications/travel-pay/components/TravelClaimDetails.jsx:
```diff
@@ -0,0 +1,137 @@
+  const [claimsError, setClaimsError] = useState(null);
+  const BTSSS_PORTAL_URL = 'https://dvagov-btsss.dynamics365portals.us/';
+
+  console.log(claimDetails); // eslint-disable-line no-console
```
Why is the linter disabled here? Let's remove this to avoid console output where not necessary.

hinzed1127 replied:
This is just a code snippet I use in local dev that needs to be cleaned up

kjduensing commented on src/applications/travel-pay/components/TravelClaimDetails.jsx:
```diff
@@ -0,0 +1,137 @@
+
+  console.log(claimDetails); // eslint-disable-line no-console
+  useEffect(() => {
+    // window.scrollTo(0, 0);
+    // scrollTo('h1');
```
Remove commented code

kjduensing commented on src/applications/travel-pay/components/TravelClaimDetails.jsx:
```diff
@@ -0,0 +1,137 @@
+        </p>
+
+        <a
+          href="https://whereDoesThisGo?.com"
```
Change to the right environment depending and use this link: https://va.gov/resources/how-to-set-up-direct-deposit-for-va-travel-pay-reimbursement/

kjduensing commented on src/applications/travel-pay/components/TravelClaimDetails.jsx:
```diff
@@ -0,0 +1,137 @@
+      {/* <MhvSecondaryNav /> */}
+
+      <article className="usa-grid-full vads-u-padding-bottom--0">
+        {/* <div className="claim-details-breadcrumb-wrapper"> */}
+        {/* <va-icon className="back-arrow" icon="arrow_back" /> */}
```
Remove commented code

kjduensing commented on src/applications/travel-pay/components/TravelClaimDetails.jsx:
```diff
@@ -0,0 +1,137 @@
+        {claimDetails && (
+          <>
+            <h1>Your travel reimbursement claim for {appointmentDate}</h1>
+            <h2>Claim number: {claimNumber}</h2>
```
There was a question raised in midpoint review of "does this need to be a heading?". @micahtaylor what should we do instead here, if anything?

micahtaylor replied:
Good memory @kjduensing. It's styled as an `h2` but not and `h2`
I documented the structure in Figma [here](https://www.figma.com/design/RJ6OVm0MWD2WprVfO7E3lp/Travel-reimbursement-status-page?node-id=4060-3843&t=5TII6yjOhopnRC1I-4). Let me know if y'all can't see it.

kjduensing commented on src/applications/travel-pay/components/TravelClaimDetails.jsx:
```diff
@@ -0,0 +1,137 @@
+        >
+          Learn how to set up direct deposit for travel pay reimbursement
+        </a>
+        <va-need-help>
+          <div slot="content">
+            <p>
+              To manage your travel claims, file a new claim, or learn what your
+              claim status means, go to our{' '}
+              <a className="btsss-portal-link" href={BTSSS_PORTAL_URL}>
+                Beneficiary Travel Self Service System (BTSSS) portal (opens in
+                new tab)
+              </a>
+              .<br />
+              Or call <va-telephone contact="8555747292" /> from 7 a.m. to 7
+              p.m. Monday through Friday. Have your claim number ready to share
+              when you call.
+            </p>
+          </div>
+        </va-need-help>
```
We should pull this out into its own component. 

1. We've used it a couple times now
2. The portal (for Veterans) will eventually be replaced

hinzed1127 replied:
I was planning on doing that as part of clean-up, but also:

> We've used it a couple times now

This is not 100% true. There's a few discrepancies between this and the same section on the other page. This might just need to be combined, but something to validate with @micahtaylor et al. To my reading, it seems like it could be a single component with the same language

kjduensing commented on src/applications/travel-pay/components/Breadcrumbs.jsx:
```diff
@@ -1,28 +1,71 @@
+      label: 'My HealtheVet',
+    },
+    {
+      href: '/',
```
This should point to the actual status page instead of root

hinzed1127 replied:
I need to double check this one more time, but this is a quirk of React Router and how it functions. This `/` is in relation to the base url of the app, i.e., `my-health/travel-claim-status` and not the top level of the site.

kjduensing replied:
Hmm, we should look into this. When I ran this locally and clicked on this breadcrumb, it took me back to the homepage (localhost:3001).

kjduensing replied:
Not to mention it's the same as the "VA.gov home" block on line 12

hinzed1127 replied:
I think you might have tried this when I was earlier along and there was (I think) another link in the breadcrumbs array, which was why I wasn't seeing the same issue later.

If it's the last link, `VaBreadcrumbs` will just append "#content" to the current url ([ref](https://github.com/department-of-veterans-affairs/component-library/blob/main/packages/web-components/src/components/va-breadcrumbs/va-breadcrumbs.tsx#L334-L343)), which was what I was seeing.

In either case, the second "/" `href` needed to be updated

kjduensing commented on src/applications/travel-pay/sass/travel-pay.scss:
```diff
@@ -52,3 +53,20 @@ h2#pagination-info {
   font-weight: 700;
 }
+
+// .claim-details-breadcrumb-wrapper {
+//   display: flex;
+
+//   align-items: center;
+
+//   .back-arrow {
+//     color: var(--vads-color-gray-medium);
+//   }
+
+//   va-breadcrumbs nav {
+//     padding: 0;
+//   }
+//   // nav {
+//   //   padding: 0 !important;
+//   // }
+// }
```
can remove?

kjduensing commented on src/applications/travel-pay/routes.jsx:
```diff
@@ -1,11 +1,17 @@
+    <Route exact path="/" title="TravelPayHome">
+      <TravelPayStatusApp />
+    </Route>
+    <Route exact path="/:id">
+      <TravelClaimDetails />
+    </Route>
```
If we can, let's put this route behind the details feature flag.

hinzed1127 replied:
Good catch. On the off chance someone tries to navigate directly to a details route, I think I'll initially throw that check in `TravelClaimDetails.jsx`. That way they can still get redirected to the status page if they have access to that but not details.

hinzed1127 commented on src/applications/travel-pay/components/Breadcrumbs.jsx:
```diff
@@ -1,28 +1,71 @@
+    },
+  ];
+
+  if (isStatusText) {
+    breadcrumbList.push({
+      href: '/what-does-my-status-mean',
+      label: 'Claim Status Meanings',
+      isRouterLink: true,
+    });
+  }
```
All of the routing updates were ripped from my earlier work on status page and modified from there. This whole PR still needs clean-up. Drawback to finishing up EOD Friday and opening a draft PR 😅

hinzed1127 commented on src/applications/travel-pay/containers/TravelPayStatusApp.jsx:
```diff
@@ -160,11 +160,11 @@ export default function App({ children }) {
   useEffect(
     () => {
-      if (userLoggedIn) {
+      if (userLoggedIn && travelClaims.length === 0) {
```
Singe we're using React Router and getting the single-page application UI benefits of that, if someone navigates to a claim details page and _back_ to this page, the current state will already have claims and so another fetch would be redundant.

I don't think I'm missing any edge cases in that logic, but I'll double check and add a clarifying comment in either case

kjduensing replied:
I think I'm missing something, probably lacking React knowledge. My question was about the first time someone comes to this page - the first time the `useEffect` fires won't it check for `travelClaims.length === 0`? I can see that this is working locally, so maybe you can let me know what I'm missing offline

hinzed1127 replied:
> My question was about the first time someone comes to this page - the first time the useEffect fires won't it check for travelClaims.length === 0?

Yes, which is the only time we want claims fetching to fire.

> won't travelClaims always have a length of 0?

Client-side routing means this isn't always true. For instance, in this scenario, it wouldn't:

1. Claims status page loads. `travelClaims.length` is 0, so claims get fetch
2. User then navigates to a claim details page. Since right now this is just displaying info that's exclusively found in the claim data from the status page, there's no additional fetch for the claim details
3. User navigates from claim details **_back_** to claim status page using the breadcrumb navigation. Since this uses client-side routing, the previously fetched `travelClaims` are still in the state, and so re-fetching the claims is not needed

hinzed1127 committed:
Consolidate repetitive help text into functional component

hinzed1127 committed:
Add feature flag checks for claim details

hinzed1127 committed:
Lots of PR fixes

- Added in feature flag checks
- Updated the breadcrumbs on claims details page
- Added MHV nav bar across pages

hinzed1127 committed:
Remove comments

hinzed1127 committed:
Breadcrumbs cleanup

hinzed1127 committed:
HTML/CSS updates

hinzed1127 committed:
Update spacing on claim details <p>s

hinzed1127 committed:
Fix mockdata discrepancy

hinzed1127 committed:
Fix breadcrumb href

hinzed1127 committed:
Add claim details feature flag tests on status page

hinzed1127 committed:
Committing this atrocity of a test file for posterity

hinzed1127 committed:
More tests, refactor claims details card to consolidate display logic

hinzed1127 committed:
Add cypress test for claims details

kjduensing reviewed (approved):
Looks good Dan 🤖
